src/Shared/BrowserTesting/src/BrowserManager.cs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using PlaywrightSharp;
+
+namespace Microsoft.AspNetCore.BrowserTesting
+{
+    public class BrowserManager
+    {
+        private readonly BrowserManagerConfiguration _browserManagerConfiguration;
+        private readonly Dictionary<string, IBrowser> _launchBrowsers = new();
+
+        private object _lock = new();
+        private Task _initializeTask;
+        private bool _disposed;
+        private readonly ILoggerFactory _loggerFactory;
+
+        private BrowserManager(IConfiguration configuration, ILoggerFactory loggerFactory)
+        {
+            _browserManagerConfiguration = new BrowserManagerConfiguration(configuration);
+            _loggerFactory = loggerFactory;
+        }
+
+        public IPlaywright Playwright { get; private set; }
+
+        public bool HasFailedTests { get; set; }
+
+        public static async Task<BrowserManager> CreateAsync(IConfiguration configuration, ILoggerFactory loggerFactory)
+        {
+            var manager = new BrowserManager(configuration, loggerFactory);
+            await manager.InitializeAsync();
+
+            return manager;
+        }
+
+        private async Task InitializeAsync()
+        {
+            await LazyInitializer.EnsureInitialized(ref _initializeTask, ref _lock, InitializeCore);
+
+            async Task InitializeCore()
+            {
+                Playwright = await PlaywrightSharp.Playwright.CreateAsync(_loggerFactory, debug: "DEBUG=pw:api");
+                foreach (var (browserName, options) in _browserManagerConfiguration.BrowserOptions)
+                {
+                    if (!_launchBrowsers.ContainsKey(browserName))
+                    {
+                        var effectiveLaunchOptions = _browserManagerConfiguration.GetLaunchOptions(options.BrowserLaunchOptions);
+
+                        var browser = options.BrowserKind switch
+                        {
+                            BrowserKind.Chromium => await Playwright.Chromium.LaunchAsync(effectiveLaunchOptions),
```

Comment:
Does this mean Playwright runs tests in multiple browsers at the same time as opposed to doing a pass per browser?

Reply:
No, we launch the browsers ahead of time, but then we only use one at a time (only chrome for now)

For template tests I didn't want to run against firefox or safari just because it adds a significant amount of time to run `dotnet new` and so on for each browser and I don't want to try all three browsers in a single test either.